src/core/inductive.ml

```diff
@@ -0,0 +1,292 @@
+(** Generating of induction principles. *)
+
+open Timed
+open Pos
+open Console
+open Terms
+open Print
+open Syntax
+
+(** Logging function for generating of inductive principle. *)
+let log_ind = new_logger 'g' "indu" "generating induction principle"
+let log_ind = log_ind.logger
+
+(** Builtin configuration for induction. *)
+type config =
+  { symb_Prop : sym (** : TYPE. Type of propositions. *)
+  ; symb_prf  : sym (** : Prop → TYPE.
+                        Interpretation of propositions as types. *) }
+
+(** [get_config ss pos] build the configuration using [ss]. *)
+let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
+  let builtin = Builtin.get ss pos in
+  { symb_Prop = builtin "Prop"
+  ; symb_prf  = builtin "P" }
+
+(** [check_typ_ind rec_typ] checks if the type of the term [rec_typ] is the
+    constant TYPE. *)
+let check_type_ind : popt -> term -> unit = fun pos rec_typ ->
+  match Typing.infer [] rec_typ with
+  | Some t ->
+      begin
+      match unfold t with
+      | Type   -> ()
+      | _ -> fatal pos "The type of the generated inductive principle of [%a]
+                        isn't the constant TYPE. Please, raise an issue."
+               pp_term rec_typ
+      end
+  | None   ->
+      fatal pos "The type of the generated inductive principle of
+                 [%a] isn't typable. Please, raise an issue."
+        pp_term rec_typ
+
+(** [check_name_constructor i l] creates the prefixes for variables' name and
+    predicats' name according to the name of the inductive type [i] and its
+    constructors [l]. *)
+let check_name_constructor : sym -> sym list -> string * string = fun i l ->
+  (* [aux_2acc l acc1 acc2] returns the list of constructors' name which are
+     in conflict with the names of the form "x" with a number or not (stored
+     in [acc1]) or in conflict with the names of the form "p" with a number
+     or not (stored in [acc2]). *)
+  let rec aux_2acc l acc1 acc2 = match l with
+    | [] -> acc1, acc2
+    | t::q ->
+        let s = t.sym_name in
+        if Str.string_match (Str.regexp "^x[0-9]*$") s 0 then
```

Comment:
Remark: we could avoid dependency to Str by testing whether the string is not empty, test if its first character is 'x' or 'p', and use scanf.

Reply:
By using scanf, you could merge the functions check_name_constructor and get_prefix into a single function.

Reply:
It would be good also to abstract other "x" and "p", and perhaps move this to extra.ml:
```
(** [get_safe_prefix p strings] returns a string [p'] starting with [p] and so that, for no positive integer [k], [p' ^ string_of_int k] belongs to [strings].
let get_safe_prefix : string -> StrSet.t -> string = ...
```
Note that, when building the argument [strings] from a list of symbols, you can add only the names starting with "x" or "p".

Reply:
Why would you like to test if the string is not empty? Isn't possible ?

Reply:
Well, to check the first character, you first need to make sure that it exists.
Btw, we have to be careful that we may have identifiers of the form "x00". When using scanf on "00", the result is the same as with "0" or "000". But leading zeros are not important if we always add +1, right?

Reply:
Hum, in fact, your "p" is a string but it can have a length > 1, no?
Instead of using "scanf", I used only "string_of_int" with a "try ... with ..." structure. See on Extra.ml!